pkg/cloudprovider/servicecontroller/servicecontroller.go

```diff
@@ -0,0 +1,426 @@
+			break
+		}
+	} else {
+		endpoint, err := s.balancer.CreateTCPLoadBalancer(name, s.zone.Region, nil,
+			ports, hostsFromNodeList(nodes), service.Spec.SessionAffinity)
+		if err != nil {
+			return err
+		}
+		service.Spec.PublicIPs = []string{endpoint}
+	}
+	return nil
+}
+
+func (s *ServiceController) handleDelete(service *api.Service) error {
+	if err := s.deleteExternalLoadBalancer(service); err != nil {
+		return err
+	}
+	namespacedName, err := cache.MetaNamespaceKeyFunc(service)
+	if err != nil {
+		// This is panic-worthy, since the queue shouldn't have been able to
+		// handle the service if it couldn't generate a name for it.
+		glog.Errorf("Couldn't generate namespaced name for service: %v", err)
+		return nil
+	}
+	s.deleteService(namespacedName)
+	return nil
+}
+
+func (s *ServiceController) deleteExternalLoadBalancer(service *api.Service) error {
+	// This is only needed because not all delete load balancer implementations
+	// are currently idempotent to the LB not existing.
+	if exists, err := s.balancer.TCPLoadBalancerExists(s.loadBalancerName(service), s.zone.Region); err != nil {
+		return err
+	} else if !exists {
+		return nil
+	}
+
+	if err := s.balancer.DeleteTCPLoadBalancer(s.loadBalancerName(service), s.zone.Region); err != nil {
+		return err
+	}
+	return nil
+}
+
+// listKeys implements the interface required by DeltaFIFO to list the keys we
+// already know about.
+func (s *ServiceController) listKeys() []string {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	keys := make([]string, len(s.serviceMap))
+	i := 0
+	for k := range s.serviceMap {
+		keys[i] = k
+		i++
+	}
+	return keys
+}
+
+func (s *ServiceController) getService(serviceName string) (*api.Service, bool) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	info, ok := s.serviceMap[serviceName]
+	return info, ok
+}
+
+func (s *ServiceController) setService(serviceName string, info *api.Service) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.serviceMap[serviceName] = info
+}
+
+func (s *ServiceController) deleteService(serviceName string) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	delete(s.serviceMap, serviceName)
+}
+
+func needsUpdate(oldService *api.Service, newService *api.Service) bool {
+	if !oldService.Spec.CreateExternalLoadBalancer && !newService.Spec.CreateExternalLoadBalancer {
+		return false
+	}
+	if oldService.Spec.CreateExternalLoadBalancer != newService.Spec.CreateExternalLoadBalancer {
+		return true
+	}
+	if !portsEqual(oldService, newService) || oldService.Spec.SessionAffinity != newService.Spec.SessionAffinity {
+		return true
+	}
+	if len(oldService.Spec.PublicIPs) != len(newService.Spec.PublicIPs) {
+		return true
+	}
+	for i := range oldService.Spec.PublicIPs {
+		if oldService.Spec.PublicIPs[i] != newService.Spec.PublicIPs[i] {
+			return true
+		}
+	}
+	return false
+}
+
+func (s *ServiceController) loadBalancerName(service *api.Service) string {
+	return s.clusterName + "-" + service.Namespace + "-" + service.Name
```

Comment:
I noticed recently that this can quite exceed the name length limits on GCE, and probably other cloud providers too.  Consider using a shorter hash instead, that's crafted to fit within the limits?